### PR TITLE
Added AND search and tests

### PR DESCRIPTION
**Pull request by brianignacio5**
#27 Please do a code review and merge if it's ok.

### CODE REVIEW

**Comment by brianignacio5** [comment]
https://github.com/moritzschaefer/the-search-engine/issues/27

**Comment by brianignacio5** [comment]
#27

File: test/test_ranker.py [hunk 1]
```diff
@@ -67,30 +67,63 @@ def test_term_query_weight():
 
     eq_(round(term_weight,4), round(weight,4))
 
+@with_setup(create_dictionary_index, remove_test_dict_info)
+def test_conjunctive_score_calc():
+    query_terms = ['term', 'to', 'evaluate']
+
+    scores = {'15da4df3-9ef1-4e1a-b0ba-f93bf05a25d0': 0.9998169851514684,
```
Comment (moritzschaefer) [inline test/test_ranker.py]: isn't it a bit weird that they are almost 1?

File: test/test_ranker.py [hunk 2]
```diff
@@ -67,30 +67,63 @@ def test_term_query_weight():
...
-    scores = [('15da4df3-9ef1-4e1a-b0ba-f93bf05a25d0', 0.9999986361252825),
-                ('7dd5a186-1dfe-4be6-be0b-ded65e8067c9', 0.9999986361252825),
-                ('c7c1d354-4b85-438b-bb2e-89350e40e33f', 0.9999986361252825)]
+    scores = [('7dd5a186-1dfe-4be6-be0b-ded65e8067c9', 0.9997654994406655), 
+              ('15da4df3-9ef1-4e1a-b0ba-f93bf05a25d0', 0.9998169851514684), 
+              ('c7c1d354-4b85-438b-bb2e-89350e40e33f', 0.9999984674316662)]
 
     scores_by_function = rank(query_terms, TEST_DICT_PATH, TEST_INDEXINFO_PATH)
 
+    print(scores_by_function)
```
Comment (moritzschaefer) [inline test/test_ranker.py]: Delete this please

File: test/test_ranker.py [hunk 3]
```diff
@@ -67,30 +67,63 @@ def test_term_query_weight():
 
     eq_(round(term_weight,4), round(weight,4))
 
+@with_setup(create_dictionary_index, remove_test_dict_info)
+def test_conjunctive_score_calc():
+    query_terms = ['term', 'to', 'evaluate']
+
+    scores = {'15da4df3-9ef1-4e1a-b0ba-f93bf05a25d0': 0.9998169851514684, 
+              '7dd5a186-1dfe-4be6-be0b-ded65e8067c9': 0.9997654994406655}
```
Comment (moritzschaefer) [inline test/test_ranker.py]: why is there 1 score missing here? the others have 3..

File: tsg/ranker/base.py [hunk 4]
```diff
@@ -64,11 +105,11 @@ def cosine_score_calc(query_terms, index_dictionary_path=DICTIONARY_PATH,
                 scores[key] = float(value)*float(query_term_weight)
 
             if key in query_length:
-                query_length[key] += math.pow(float(query_term_weight),float(2))
```
Comment (moritzschaefer) [inline tsg/ranker/base.py]: good :)

File: tsg/ranker/base.py [hunk 5]
```diff
@@ -78,19 +119,29 @@ def cosine_score_calc(query_terms, index_dictionary_path=DICTIONARY_PATH,
 
     return scores
 
+def combine_and_or_scores(and_dict, or_dict):
+    sorted_and = sorted(and_dict.items(), key = operator.itemgetter(1,0))
+    sorted_or = sorted(or_dict.items(), key = operator.itemgetter(1, 0))
+
+    new_docs = set(sorted_or) - set(sorted_and)
```
Comment (moritzschaefer) [inline tsg/ranker/base.py]: This should not be working. you are comparing 2-tuples with weight which are always different between or_list and and_list. am I confused?

File: tsg/ranker/base.py [hunk 6]
```diff
@@ -47,6 +47,47 @@ def calculate_query_term_weight(term, query_terms, index_dictionary_path=DICTION
...
+    index_info_path = INDEXINFO_PATH):
+    
+    results_keys = set()
+    terms_lists = {}
+    scores = {}
+    doc_length = {}
+    query_length = {}
+
+    for term in query_terms:
+        terms_lists[term] = get_dictionary_term_list(term, index_dictionary_path)
```
Comment (moritzschaefer) [inline tsg/ranker/base.py]: this is STILL not a terms_list. its a list with document

File: test/test_ranker.py (hunk 1 above)
Comment (brianignacio5) [inline test/test_ranker.py]: Is difficult to actually know the exact value. Conjunctive search is correct name for AND search.

File: test/test_ranker.py (hunk 3 above)
Comment (brianignacio5) [inline test/test_ranker.py]: Since this is AND search, we only get results that includes all query terms.

File: test/test_ranker.py (hunk 2 above)
Comment (brianignacio5) [inline test/test_ranker.py]: Missed this.

File: tsg/ranker/base.py (hunk 5 above)
Comment (brianignacio5) [inline tsg/ranker/base.py]: You are confused. Here we are getting those documents that belongs to Sorted_or  but don't belong to sorted_and. The idea is to get those exact match (docs in AND) first then we add the new_docs.

File: tsg/ranker/base.py (hunk 6 above)
Comment (brianignacio5) [inline tsg/ranker/base.py]: Ok, I'll just rename it with doc_list_per_term.

Comment (moritzschaefer) [inline tsg/ranker/base.py]: doc_list_per_term. sounds horrible.
term_documents is okay. If you want doc_list_FOR_term is also ok

File: tsg/ranker/base.py [hunk 7]
```diff
@@ -78,19 +119,29 @@ def cosine_score_calc(query_terms, index_dictionary_path=DICTIONARY_PATH,
...
+
+    return scores
+
 def rank(query_terms, index_dictionary_path=DICTIONARY_PATH,
     index_info_path = INDEXINFO_PATH):
     '''
     Ranker takes a query and a dictionary path to calculates the score
     and retrieved a list of docs ordered by score and doc_id as
     tuples [(doc_1,score_1),(doc_2, score_2), ..., (doc_n,score_n)]
     '''
```
Comment (moritzschaefer) [inline tsg/ranker/base.py]: And the idea was not, that we do conjunctive and disjunctive and then just combine them...
some terms may be conjunctive, some disjunctive.

File: tsg/ranker/base.py [hunk 8]
```diff
@@ -47,62 +47,62 @@ def calculate_query_term_weight(term, query_terms, index_dictionary_path=DICTION
...
     return weight
 
-def conjunctive_score_calc(query_terms, index_dictionary_path= DICTIONARY_PATH, 
+def and_score_calc(query_terms, index_dictionary_path= DICTIONARY_PATH, 
     index_info_path = INDEXINFO_PATH):
     
-    results_keys = set()
-    terms_lists = {}
-    scores = {}
+    common__doc_keys = set()
```
Comment (moritzschaefer) [inline tsg/ranker/base.py]: double underscore? ;)

**Pull request merged**